nmtwizard/prepoperator.py

```diff
@@ -62,24 +63,201 @@ def __call__(self):
+
+    def __init__(self, config, result_dir):
+
+        super(VocabularyBuilder, self).__init__(result_dir)
 
-    def __init__(self, f, preprocess_dir):
+        self._vocabularies = {}
+        self._sums = {}
 
-        self._preprocess_dir = preprocess_dir
+        opt_multi = config.get('tokenization', {}).get('multi')
+        opt_source = config.get('tokenization', {}).get('source')
+        opt_target = config.get('tokenization', {}).get('target')
 
-        # TODO : multiple files
-        # TODO : do we output ALL the files that we take as input ?
-        src = os.path.join(preprocess_dir, os.path.basename(f.files[0].name))
+        if opt_multi:
+            self._vocabularies['multi'] = {}
+            self._sums['multi'] = 0
+        if opt_source:
+            self._vocabularies['source'] = {}
+            self._sums['source'] = 0
+        if opt_target:
+            self._vocabularies['target'] = {}
+            self._sums['target'] = 0
+
+
+    def __call__(self, tu_batch):
+
+        # TODO V2 : feed tokenized words ?
+        # TODO : remove value for placeholders
+        for tu in tu_batch :
+            if 'source' in self._vocabularies:
+                for token in tu.src_raw.split():
+                    self._vocabularies['source'][token] = \
+                        self._vocabularies['source'].get(token, 0) + 1
+                    self._sums['source'] += 1
+            if 'target' in self._vocabularies:
+                for token in tu.tgt_raw.split():
+                    self._vocabularies['target'][token] = \
+                        self._vocabularies['target'].get(token, 0) + 1
+                    self._sums['target'] += 1
+            if 'multi' in self._vocabularies:
+                for token in tu.src_raw.split():
+                    self._vocabularies['multi'][token] = \
+                        self._vocabularies['multi'].get(token, 0) + 1
+                    self._sums['multi'] += 1
+                for token in tu.tgt_raw.split():
+                    self._vocabularies['multi'][token] = \
+                        self._vocabularies['multi'].get(token, 0) + 1
+                    self._sums['multi'] += 1
+
+    def finalize(self, config):
+
+        for side, vocabulary in self._vocabularies.items():
+            name =  config['tokenization'][side]['vocabulary']['name'] \
+                    if 'name' in config['tokenization'][side]['vocabulary'] \
+                    else 'vocab'
+
+            # Size option is mandatory, already checked it.
+            size = config['tokenization'][side]['vocabulary']['size']
+
+            min_frequency = config['tokenization'][side]['vocabulary']['min-frequency'] \
+                            if 'min-frequency' in config['tokenization'][side]['vocabulary'] \
+                            else 0
+
+            # Find out the real vocabulary size.
+            sorted_tokens = sorted(vocabulary, key=vocabulary.get, reverse=True)
```

Comment:
The logic of pruning the vocabulary by size and frequency should be moved to a separate function and unit tested.